File: Skribe/Skribe/Services/VADProcessor.swift
import Foundation
import Accelerate

/// Simple energy-based VAD for skipping silence
/// Reduces processing time by ~20-30% on typical meeting audio
struct VADProcessor {

    struct VoiceSegment {
        let startSample: Int
        let endSample: Int
        let startTime: Double
        let endTime: Double
    }

    /// Detect voice activity in audio samples
    /// - Parameters:
    ///   - samples: Audio samples (16kHz mono)
    ///   - sampleRate: Sample rate (default 16000)
    ///   - frameSize: Analysis frame size in samples (default 400 = 25ms)
    ///   - hopSize: Hop size in samples (default 160 = 10ms)
    ///   - energyThreshold: Energy threshold (relative to max, 0.0-1.0, default 0.02)
    ///   - minSilenceDuration: Minimum silence duration to split segments (default 0.3s)
    ///   - minSpeechDuration: Minimum speech duration to keep (default 0.1s)
    /// - Returns: Array of voice segments
    static func detectVoiceActivity(
        samples: [Float],
        sampleRate: Int = 16000,
        frameSize: Int = 400,
        hopSize: Int = 160,
        energyThreshold: Float = 0.02,
        minSilenceDuration: Double = 0.3,
        minSpeechDuration: Double = 0.1
    ) -> [VoiceSegment] {
        guard !samples.isEmpty else { return [] }

        // Calculate frame energy
        var energies: [Float] = []
        var maxEnergy: Float = 0

        var frameStart = 0
        while frameStart + frameSize <= samples.count {
            let frameEnd = min(frameStart + frameSize, samples.count)
            let frame = Array(samples[frameStart..<frameEnd])

            // Calculate RMS energy
            var energy: Float = 0
            vDSP_measqv(frame, 1, &energy, vDSP_Length(frame.count))
            energy = sqrt(energy / Float(frame.count))

            energies.append(energy)
            maxEnergy = max(maxEnergy, energy)

            frameStart += hopSize
        }

        guard maxEnergy > 0 else { return [] }

        // Normalize energies and apply threshold
        let absoluteThreshold = maxEnergy * energyThreshold
        let voiceFrames = energies.map { $0 >= absoluteThreshold }

        // Find voice segments
        var segments: [VoiceSegment] = []
        var inVoice = false
        var segmentStart = 0

        let minSilenceFrames = Int(minSilenceDuration * Double(sampleRate) / Double(hopSize))
        let minSpeechFrames = Int(minSpeechDuration * Double(sampleRate) / Double(hopSize))
        var silenceCounter = 0

        for (i, isVoice) in voiceFrames.enumerated() {
            if isVoice {
                if !inVoice {
                    // Start new voice segment
                    segmentStart = i
                    inVoice = true
                    silenceCounter = 0
                }
            } else {
                if inVoice {
                    silenceCounter += 1
                    if silenceCounter >= minSilenceFrames {
                        // End voice segment
                        let segmentEnd = i - silenceCounter
                        let duration = segmentEnd - segmentStart

                        if duration >= minSpeechFrames {
                            let startSample = segmentStart * hopSize
                            let endSample = min(segmentEnd * hopSize, samples.count)

                            segments.append(VoiceSegment(
                                startSample: startSample,
                                endSample: endSample,
                                startTime: Double(startSample) / Double(sampleRate),
                                endTime: Double(endSample) / Double(sampleRate)
                            ))
                        }

                        inVoice = false
                        silenceCounter = 0
                    }
                }
            }
        }

        // Handle final segment
        if inVoice {
            let segmentEnd = voiceFrames.count
            let duration = segmentEnd - segmentStart

            if duration >= minSpeechFrames {
                let startSample = segmentStart * hopSize
                let endSample = samples.count

                segments.append(VoiceSegment(
                    startSample: startSample,
                    endSample: endSample,
                    startTime: Double(startSample) / Double(sampleRate),
                    endTime: Double(endSample) / Double(sampleRate)
                ))
            }
        }

        // Merge close segments (within 0.5s)
        return mergeCloseSegments(segments, maxGap: 0.5)
    }

    /// Merge segments that are close together
    private static func mergeCloseSegments(_ segments: [VoiceSegment], maxGap: Double) -> [VoiceSegment] {
        guard segments.count > 1 else { return segments }

        var merged: [VoiceSegment] = []
        var current = segments[0]

        for i in 1..<segments.count {
            let next = segments[i]
            let gap = next.startTime - current.endTime

            if gap <= maxGap {
                // Merge with current
                current = VoiceSegment(
                    startSample: current.startSample,
                    endSample: next.endSample,
                    startTime: current.startTime,
                    endTime: next.endTime
                )
            } else {
                // Save current and start new
                merged.append(current)
                current = next
            }
        }

        merged.append(current)
        return merged
    }

    /// Extract only voice segments from audio samples
    static func extractVoiceSegments(samples: [Float], voiceSegments: [VoiceSegment]) -> [Float] {
        var result: [Float] = []

        for segment in voiceSegments {
            let segmentSamples = Array(samples[segment.startSample..<segment.endSample])
            result.append(contentsOf: segmentSamples)
        }

        return result
    }
}

File: windows/Skribe/Services/VADProcessor.cs
using System;
using System.Collections.Generic;

namespace Skribe.Services;

/// <summary>
/// Lightweight energy-based VAD. Mirrors the Swift implementation:
/// produces speech segments to avoid wasting whisper time on silence.
/// </summary>
public static class VADProcessor
{
    public sealed record VoiceSegment(double StartTime, double EndTime);

    private const int SampleRate = 16_000;
    private const int FrameMs = 30;
    private const int FrameSamples = SampleRate * FrameMs / 1000; // 480
    private const float EnergyThreshold = 0.005f;
    private const int MinSpeechFrames = 5;   // ~150 ms
    private const int MinSilenceFrames = 20; // ~600 ms

    public static List<VoiceSegment> DetectVoiceActivity(float[] samples)
    {
        var result = new List<VoiceSegment>();
        if (samples.Length == 0) return result;

        bool inSpeech = false;
        int speechFrameCount = 0;
        int silenceFrameCount = 0;
        int segmentStart = 0;

        for (int frameStart = 0; frameStart + FrameSamples <= samples.Length; frameStart += FrameSamples)
        {
            double energy = 0;
            for (int i = 0; i < FrameSamples; i++)
            {
                var s = samples[frameStart + i];
                energy += s * s;
            }
            energy = Math.Sqrt(energy / FrameSamples);

            bool isSpeech = energy > EnergyThreshold;

            if (isSpeech)
            {
                if (!inSpeech)
                {
                    speechFrameCount++;
                    if (speechFrameCount >= MinSpeechFrames)
                    {
                        inSpeech = true;
                        segmentStart = frameStart - (speechFrameCount - 1) * FrameSamples;
                        silenceFrameCount = 0;
                    }
                }
                else
                {
                    silenceFrameCount = 0;
                }
            }
            else if (inSpeech)
            {
                silenceFrameCount++;
                if (silenceFrameCount >= MinSilenceFrames)
                {
                    int segEnd = frameStart - silenceFrameCount * FrameSamples;
                    result.Add(new VoiceSegment(
                        (double)segmentStart / SampleRate,
                        (double)segEnd / SampleRate));
                    inSpeech = false;
                    speechFrameCount = 0;
                    silenceFrameCount = 0;
                }
            }
            else
            {
                speechFrameCount = 0;
            }
        }

        if (inSpeech)
        {
            result.Add(new VoiceSegment(
                (double)segmentStart / SampleRate,
                (double)samples.Length / SampleRate));
        }

        return result;
    }
}
